drivers/source/UARTSerial.cpp

```diff
@@ -366,38 +366,18 @@ void UARTSerial::disable_tx_irq()
 
 int UARTSerial::enable_input(bool enabled)
 {
-    core_util_critical_section_enter();
-    if (_rx_enabled != enabled) {
-        if (enabled) {
-            UARTSerial::rx_irq();
-            if (!_rxbuf.full()) {
-                enable_rx_irq();
-            }
-        } else {
-            disable_rx_irq();
-        }
-        _rx_enabled = enabled;
-    }
-    core_util_critical_section_exit();
+    api_lock();
+    SerialBase::enable_input(enabled);
```

Comment:
Where is the call to `enable_rx_irq` to attach the rx Irq?

Reply:
The interrupt handler remains attached. It activated by enabling the rx irq in `SerialBase::enable_input`.

But i think there is a real problem with this version, as the device is not allowed to go to deep sleep, because deep sleep is unblocked by passing `nullptr` as func to `SerialBase::attach` (which is not called).

Reply:
Good catch;  the `SerialBase::enable_input`/`output` functions will have to manipulate the locks. Not sure what the most efficient way of handling that is - ideally you'd be able to have some common "update HAL IRQ + lock state" call that was used by both `attach` and `enable`.